tests/api2/test_smb_null_empty_dacl.py

```diff
@@ -0,0 +1,135 @@
+        case _:
+            raise TypeError(f'[EDOOFUS]: {special_acl_type} unexpected special ACL type')
+
+    payload = json.dumps({'acl': [{
+        'tag': 'everyone@',
+        'id': -1,
+        'type': 'ALLOW',
+        'perms': permset,
+        'flags': {'BASIC': 'NOINHERIT'},
+    }]})
+    ssh(f'touch {path}')
+
+    # Use SSH to write to avoid middleware ACL normalization and validation
+    # that prevents writing these specific ACLs.
+    ssh(f"nfs4xdr_setfacl -j '{payload}' {path}")
+
+
+def test_null_dacl_set(unprivileged_user_fixture, share):
+    """ verify that setting NULL DACL results in expected ZFS ACL """
+    with smb_connection(
+        share=share['share']['name'],
+        username=unprivileged_user_fixture.username,
+        password=unprivileged_user_fixture.password,
+    ) as c:
+        fh = c.create_file('test_null_dacl', 'w')
+        current_sd = c.get_sd(fh, security.SECINFO_OWNER | security.SECINFO_GROUP)
+        current_sd.dacl = None
+        c.set_sd(fh, current_sd, security.SECINFO_OWNER | security.SECINFO_GROUP | security.SECINFO_DACL)
+
+        new_sd = c.get_sd(fh, security.SECINFO_OWNER | security.SECINFO_GROUP)
+        assert new_sd.dacl is None
+
+        theacl = call('filesystem.getacl', os.path.join(share['share']['path'], 'test_null_dacl'))
+        assert len(theacl['acl']) == 1
+
+        assert theacl['acl'][0]['perms'] == NULL_DACL_PERMS
+        assert theacl['acl'][0]['type'] == 'ALLOW'
+        assert theacl['acl'][0]['tag'] == 'everyone@'
+
+
+def test_null_dacl_functional(unprivileged_user_fixture, share):
+    """ verify that NULL DACL grants write privileges """
+    testfile = os.path.join(share['share']['path'], 'test_null_dacl_write')
+    set_special_acl(testfile, 'NULL_DACL')
+    data = b'canary'
+
+    with smb_connection(
+        share=share['share']['name'],
+        username=unprivileged_user_fixture.username,
+        password=unprivileged_user_fixture.password,
+    ) as c:
+        fh = c.create_file('test_null_dacl_write', 'w')
+        current_sd = c.get_sd(fh, security.SECINFO_OWNER | security.SECINFO_GROUP)
+        assert current_sd.dacl is None
+
+        c.write(fh, data)
+        assert c.read(fh, 0, cnt=len(data)) == data
+
+
+def test_empty_dacl_set(unprivileged_user_fixture, share):
+    """ verify that setting empty DACL results in expected ZFS ACL """
+    with smb_connection(
+        share=share['share']['name'],
+        username=unprivileged_user_fixture.username,
+        password=unprivileged_user_fixture.password,
+    ) as c:
+        fh = c.create_file('test_null_dacl', 'w')
+        current_sd = c.get_sd(fh, security.SECINFO_OWNER | security.SECINFO_GROUP)
+        current_sd.dacl = security.acl()
+        c.set_sd(fh, current_sd, security.SECINFO_OWNER | security.SECINFO_GROUP | security.SECINFO_DACL)
+
+        new_sd = c.get_sd(fh, security.SECINFO_OWNER | security.SECINFO_GROUP)
+        assert new_sd.dacl.num_aces == 0
```

Comment:
If this _**is**_ an empty DACL test, then how would we have permissions to read this thru SMB?

Reply:
This is why the next test verifies that non-owner gets access denied.